feat: [FC-86] add toggle functionallity for new catalog MFE

andrii-hantkovskyi opened the pull request:
### Enable New Catalog MFE via Waffle and Feature Flag

This PR introduces two toggles to support a staged rollout of the new Catalog micro-frontend (MFE):

- `FEATURES['ENABLE_CATALOG_MICROFRONTEND']`: A Django setting for platform-level control (with SiteConfiguration override).

The `use_catalog_mfe()` helper checks the feature flag. A unit test ensures the waffle flag logic works as expected.

andrii-hantkovskyi committed:
feat: [AXM-2056] add catalog MFE enabled waffle flag (#2639)

* feat: [AXM-2056] add catalog MFE enabled waffle flag

* test: cover waffle with test

* refactor: fix pylint error

* refactor: redesign waffleflag & add new MFE toggle to FEATURES

* test: update tests to new functionality

* refactor: add toggle_use_cases to annotation

* refactor: redesign again to use new approach

* refactor: rename waffle flag variable & update description

---------

Co-authored-by: Andrii <andrii.hantkovskyi@raccoongang.com>

openedx-webhooks commented:
Thanks for the pull request, @andrii-hantkovskyi!


This repository is currently maintained by `@openedx/wg-maintenance-edx-platform`.


Once you've gone through the following steps feel free to tag them in a comment and let them know that your changes are ready for engineering review.

<details><summary>:radio_button: Get product approval</summary>

If you haven't already, [check this list](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Does-my-contribution-require-Product-Review%3F) to see if your contribution needs to go through the product review process.

- If it does, you'll need to submit a product proposal for your contribution, and have it reviewed by the [Product Working Group](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3449028609/Product+Working+Group).
    - This process (including the steps you'll need to take) is documented [here](https://openedx.atlassian.net/wiki/spaces/COMM/pages/3875962884/How+to+submit+an+open+source+contribution+for+Product+Review#Product-Review-Process).
- If it doesn't, simply proceed with the next step.
</details>

<details><summary>:radio_button: Provide context</summary>

To help your reviewers and other members of the community understand the purpose and larger context of your changes, feel free to add as much of the following information to the PR description as you can:

- Dependencies
  > This PR must be merged before / after / at the same time as ...
- Blockers
  > This PR is waiting for OEP-1234 to be accepted.
- Timeline information
  > This PR must be merged by XX date **because** ...
- Partner information
  > This is for a course on edx.org.
- Supporting documentation
- Relevant [Open edX discussion forum](https://discuss.openedx.org/) threads
</details>





<details><summary>:radio_button: Get a green build</summary>

If one or more checks are failing, continue working on your changes until this is no longer the case and your build turns green.
</details><details>


---

<details><summary>Where can I find more information?</summary>

If you'd like to get more details on all aspects of the review process for open source pull requests (OSPRs), check out the following resources:

- [Overview of Review Process for Community Contributions](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/FAQ-about-pull-requests.html)
- [Pull Request Status Guide](https://docs.openedx.org/en/latest/developers/references/developer_guide/process/pull-request-statuses.html)
- [Making changes to your pull request](https://docs.openedx.org/en/latest/documentors/how-tos/make_changes_to_your_pull_request.html)
</details>

<details><summary>When can I expect my changes to be merged?</summary>

Our goal is to get community contributions seen and reviewed as efficiently as possible.

However, the amount of time that it takes to review and merge a PR can vary significantly based on factors such as:

- The size and impact of the changes that it introduces
- The need for product review
- Maintenance status of the parent repository

:bulb: *As a result it may take up to several weeks or months to complete a review and merge your PR.*
</details>
<!-- comment:external_pr -->
<!-- data: eyJkcmFmdCI6IGZhbHNlfQ== -->

andrii-hantkovskyi commented:
@diana-villalvazo-wgu, whenever you have some time, could you please take a look at this PR?

diana-villalvazo-wgu reviewed (approved):
LGTM

sarina commented on lms/envs/common.py:
```diff
@@ -881,6 +881,15 @@
+    # .. toggle_description: Supports staged rollout of a new micro-frontend-based implementation of the catalog.
+    # .. toggle_use_cases: temporary, open_edx
+    # .. toggle_creation_date: 2025-05-15
+    # .. toggle_target_removal_date: None
```
Toggle target removal date is not None. We will be removing this after Ulmo is cut, so target removal date should be like 2025-11-01. Thoughts @brian-smith-tcril ?

andrii-hantkovskyi replied:
added

sarina commented on lms/djangoapps/branding/toggles.py:
```diff
@@ -0,0 +1,36 @@
+# .. toggle_default: False
+# .. toggle_description: Set to True to enable the new catalog page.
+# .. toggle_creation_date: 2025-05-15
+# .. toggle_target_removal_date: None
```
This should also not be None

andrii-hantkovskyi replied:
added

Commit:
refactor: set toggle_target_removal_date

sarina commented:
@andrii-hantkovskyi this PR should add the flag in 3 places, yes? For all 3 pages (index, about page, catalog page)?

andrii-hantkovskyi commented:
@sarina, 2 other PRs added those waffla-flags, which you closed

But I think I can add those changes here, if they're needed in the result

brian-smith-tcril commented:
The two other PRs were
* https://github.com/openedx/edx-platform/pull/36843
* https://github.com/openedx/edx-platform/pull/36844

I don't see anything in either of those PRs that actually _uses_ the new waffle flags to change functionality, so I'm assuming the original plan was to eventually add logic that checks `use_new_index_page()` somewhere to redirect to the new MFE index page and `use_new_course_about_page()` somewhere to redirect to the new MFE course about page.

This PR only has `use_new_catalog_page()`, so I think the new plan is to:
* Rename the `use_new_catalog_page()` helper to something more generic (maybe `use_catalog_mfe()`?)
* Use that helper instead of `use_new_index_page()` and `use_new_course_about_page()` in any future redirect logic.

Commit:
refactor: remove per-page waffle flag in favour of global one

andrii-hantkovskyi commented:
Removed `use_new_catalog_page()` (this was also a per-page flag) and renamed `catalog_mfe_enabled()` to `use_catalog_mfe()`

When this PR is merged, I'll update dependent to use that helper

ormsbee commented on lms/envs/common.py:
```diff
@@ -881,6 +881,15 @@
+    # .. toggle_use_cases: temporary, open_edx
+    # .. toggle_creation_date: 2025-05-15
+    # .. toggle_target_removal_date: 2025-11-01
+    'ENABLE_CATALOG_MICROFRONTEND': os.environ.get("EDXAPP_ENABLE_CATALOG_MFE", False),
```
@feanil, @kdmccormick: Is it an okay practice to bring in the defaults for a feature toggle via env var check?

kdmccormick replied:
No, we shouldn't be adding new environment variables unless necessary. Simply defaulting to `False` here is appropriate.

andrii-hantkovskyi replied:
removed env variable

kdmccormick commented on lms/djangoapps/branding/toggles.py:
```diff
@@ -0,0 +1,15 @@
+    Determine if Catalog MFE is enabled, replacing student_dashboard
+    """
+    return configuration_helpers.get_value(
+        'ENABLE_CATALOG_MICROFRONTEND', settings.FEATURES.get('ENABLE_CATALOG_MICROFRONTEND')
```
```suggestion
        'ENABLE_CATALOG_MICROFRONTEND', settings.FEATURES['ENABLE_CATALOG_MICROFRONTEND']
```
`ENABLE_CATALOG_MICROFRONTEND` is guaranteed to be a key in `FEATURES` since you defined it in lms/envs/common.py. It's safe to access it directly, no need for `.get`. It's also less error-prone to access it directly, because if the common definition is removed, then we'll get a helpful `KeyError` here rather than a silent failure.

andrii-hantkovskyi replied:
refactored

kdmccormick commented on lms/envs/common.py:
```diff
@@ -881,6 +881,15 @@
+    # .. toggle_implementation: DjangoSetting
+    # .. toggle_default: False
+    # .. toggle_description: Supports staged rollout of a new micro-frontend-based implementation of the catalog.
+    # .. toggle_use_cases: temporary, open_edx
```
```suggestion
    # .. toggle_use_cases: temporary
```

kdmccormick replied:
The `open_edx` use case is somewhat meaningless since all development is now Open-edX-focused (rather than edX.org-focused). We should probably remove it from the OEP.

andrii-hantkovskyi replied:
removed

kdmccormick commented on lms/envs/common.py:
```diff
@@ -881,6 +881,15 @@
     #   toggle does not have a target removal date.
     'ENABLE_AUTHN_MICROFRONTEND': os.environ.get("EDXAPP_ENABLE_AUTHN_MFE", False),
 
+    # .. toggle_name: FEATURES['ENABLE_CATALOG_MICROFRONTEND']
```
Just confirming, you will not be adding a Waffle flag (for per-user rollout), correct? This is OK, as long as you are OK with the options being global-on and global-off.

(If you were planning to add a Waffle flag too, then I would say just add the Waffle flag and get rid of this feature flag).

andrii-hantkovskyi replied:
Yes, no Waffle flag will be added; we decided [here](https://github.com/openedx/edx-platform/pull/36734#issuecomment-2991942157) to use one global

Commit:
refactor: get rid of env variable for ENABLE_CATALOG_MICROFRONTEND & simplify use_catalog_mfe()

kdmccormick reviewed (approved):
Looks good

sarina commented:
@kdmccormick is this OK to merge?

kdmccormick commented:
Yes
